### PR TITLE
Remove redundant `uasort()`, improve `ClassDefinition` performance

### PR DESCRIPTION
**Pull request by sjokkateer**
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | no
| RFC           | no
| QA            | no

### Description
A minor optimization by removing a redundant call to `uasort()` in the `ClassDefinition::reflectParameters()` method.

To test for equality the following class and test were used:
```php
class SomeClass
{
    public function __construct(
        string $first,
        int $second,
        $third
    ) {
    }
}

final class ClaimTest extends TestCase
{
    public function testRedundantUaSortInClassDefinition(): void
    {
        $reflectionClass       = new ReflectionClass(SomeClass::class);
        $constructor           = $reflectionClass->getConstructor();
        $constructorParameters = $constructor->getParameters();

        $parameters = [];
        foreach ($constructorParameters as $parameter) {
            $parameters[$parameter->getName()] = $parameter;
        }

        // Before
        static::assertEquals(
            $constructorParameters,
            array_values($parameters)
        );

        uasort(
            $parameters,
            fn ($a, $b) => $a->getPosition() - $b->getPosition()
        );

        // After
        static::assertEquals(
            $constructorParameters,
            array_values($parameters)
        );
    }
}
```
The result of the test:
```
.                                                                   1 / 1 (100%)

Time: 00:00.002, Memory: 6.00 MB

OK (1 test, 2 assertions)
```

The `testRedundantUaSortInClassDefinition()` method demonstrates the current behaviour of the `ClassDefenition::reflectParameters()` method.

Within this method, the current class' constructor parameters are obtained through a call to the getParameters() method. According to the documentation of the [ReflectionFunctionAbstract::getParameters()](https://www.php.net/manual/en/reflectionfunctionabstract.getparameters.php#refsect1-reflectionfunctionabstract.getparameters-description) method, these parameters come in the order in which they are defined in the source. Thus, in the order in which they are defined within the class' constructor.

Within the foreach loop the parameters are added by their parameter name. PHP's [array](https://www.php.net/manual/en/language.types.array.php#language.types.array) is an *ordered map* and these will thus follow the same order in which they are defined within the class' constructor.

This means that both the original array of constructor parameters and a call to `array_values()` on the newly created parameters array result in two identical arrays. This is validated by the first assertion of the test.

Based on the comparison function that is provided to the`uasort()` we know the order will be defined by the underlying `ReflectionParameter::getPosition()` method, which sorts the array in increasing order as positions are unique. This will therefore result in an array identical to the result of the call to`ReflectionFunctionAbstract::getParameters()` validated by the last assertion of the test. It should therefore be safe to delete the `uasort()` function call.

If there is anything I misinterpreted or overlooked please let me know.

### CODE REVIEW

File: src/Definition/Reflection/ClassDefinition.php [hunk 1]
```diff
@@ -83,11 +80,6 @@ private function reflectParameters(): void
             $parameter                               = new Parameter($parameterReflection);
             $this->parameters[$parameter->getName()] = $parameter;
         }
-
-        uasort(
```
Comment (Ocramius) [inline src/Definition/Reflection/ClassDefinition.php:87]: @sjokkateer can we please make sure that your test is also committed with these changes?

**Review by Ocramius — CHANGES_REQUESTED** [review]
Tests needed to avoid going into a regression scenario (test must be green before/after).

Meanwhile, WDYT about adding a benchmark test suite with some minimal examples in it, like we do in https://github.com/laminas/laminas-servicemanager/tree/dc255c268124f19709f1494aa1259967d2ecab1a/benchmarks ?

File: src/Definition/Reflection/ClassDefinition.php (hunk 1 above)
Comment (sjokkateer) [reply src/Definition/Reflection/ClassDefinition.php:87]: Yes, does this suffice? It is however a test for demonstrative purposes though, the real assertions are done by existing tests I assume, that resolve objects.

**Comment by sjokkateer** [comment]
> Tests needed to avoid going into a regression scenario (test must be green before/after).
> 
> Meanwhile, WDYT about adding a benchmark test suite with some minimal examples in it, like we do in https://github.com/laminas/laminas-servicemanager/tree/dc255c268124f19709f1494aa1259967d2ecab1a/benchmarks ?

Is this something I could do in a separate PR or is that something you would want in this PR? Are there any guidelines I can read up on, how these work or are implemented or should I be able to derive it through the link you provided?

**Comment by sjokkateer** [comment]
> Tests needed to avoid going into a regression scenario (test must be green before/after).
> 
> Meanwhile, WDYT about adding a benchmark test suite with some minimal examples in it, like we do in https://github.com/laminas/laminas-servicemanager/tree/dc255c268124f19709f1494aa1259967d2ecab1a/benchmarks ?

I'm up for it, I'll take a closer look at the benchmark runner and the benchmark test suite soon.

File: test/ClaimTest.php [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+namespace LaminasTest\Di;
+
+use LaminasTest\Di\TestAsset\ClaimTestClass;
+use PHPUnit\Framework\TestCase;
+use ReflectionClass;
+
+use function array_values;
+use function uasort;
+
+final class ClaimTest extends TestCase
```
Comment (Ocramius) [inline test/ClaimTest.php]: This kinda work, but it should be in `ClassDefinitionTest` (if there is such test) 🤔

File: test/TestAsset/ClaimTestClass.php [hunk 3]
```diff
@@ -0,0 +1,16 @@
+<?php
+
+declare(strict_types=1);
+
+namespace LaminasTest\Di\TestAsset;
+
+class ClaimTestClass
```
Comment (Ocramius) [inline test/TestAsset/ClaimTestClass.php]: We should give this a more explicit name, perhaps explaining purpose

File: src/Definition/Reflection/ClassDefinition.php (hunk 1 above)
Comment (Ocramius) [reply src/Definition/Reflection/ClassDefinition.php:87]: This is more than enough, thanks!

**Review by Ocramius — APPROVED** [review]
:ship: lemme try before/after benchmarks before merging though :)

**Comment by Ocramius** [comment]
```
❯ ./vendor/bin/phpbench run --warmup=2 --revs=5 --iterations=10 --retry-threshold=5 --tag=before_patch_64
PHPBench (1.2.5) running benchmarks... #standwithukraine
with configuration file: /home/ocramius/Documents/laminas/laminas-di/phpbench.json
with PHP version 8.1.7, xdebug ❌, opcache ❌

\LaminasBench\Di\ClassDefinitionBench

    benchGetParametersBeforeTransparentCach.R2 I9 - Mo0.200μs (±0.00%)
    benchGetParametersBeforeTransparentCach.R1 I7 - Mo0.600μs (±0.00%)
    benchGetParametersBeforeTransparentCach.R2 I7 - Mo1.600μs (±0.00%)

Subjects: 3, Assertions: 0, Failures: 0, Errors: 0
Storing results ... OK
Run: 1348ad5397ac2b4cec5ad3197803445ec0b57c2f
```

```
❯ ./vendor/bin/phpbench run --warmup=2 --revs=5 --iterations=10 --retry-threshold=5 --tag=after_patch_64
PHPBench (1.2.5) running benchmarks... #standwithukraine
with configuration file: /home/ocramius/Documents/laminas/laminas-di/phpbench.json
with PHP version 8.1.7, xdebug ❌, opcache ❌

\LaminasBench\Di\ClassDefinitionBench

    benchGetParametersBeforeTransparentCach.R10 I9 - Mo0.000μs (±0.00%)
    benchGetParametersBeforeTransparentCach.R10 I9 - Mo0.200μs (±0.00%)
    benchGetParametersBeforeTransparentCach.R2 I2 - Mo1.000μs (±0.00%)

Subjects: 3, Assertions: 0, Failures: 0, Errors: 0

Deprecated: DOMDocument::createTextNode(): Passing null to parameter #1 ($data) of type string is deprecated in /home/ocramius/Documents/laminas/laminas-di/vendor/phpbench/dom/lib/Element.php on line 37
Storing results ... 
Deprecated: DOMDocument::createTextNode(): Passing null to parameter #1 ($data) of type string is deprecated in /home/ocramius/Documents/laminas/laminas-di/vendor/phpbench/dom/lib/Element.php on line 37
OK
Run: 1348ad529b868f7a619cc7024bacef3eb3f71848
```
```
laminas-di on  HEAD (79c991a) via 🐘 v8.1.7 via ❄️  impure (nix-shell) 
❯ ./vendor/bin/phpbench report --ref=before_patch_64 --ref=after_patch_64 --report=aggregate
+----------------------+-------------------------------------------------------------------------+-----+------+-----+----------+---------+--------+
| benchmark            | subject                                                                 | set | revs | its | mem_peak | mode    | rstdev |
+----------------------+-------------------------------------------------------------------------+-----+------+-----+----------+---------+--------+
| ClassDefinitionBench | benchGetParametersBeforeTransparentCachingNoConstructor                 |     | 5    | 10  | 1.664mb  | 0.200μs | ±0.00% |
| ClassDefinitionBench | benchGetParametersBeforeTransparentCachingNoConstructorParameters       |     | 5    | 10  | 1.664mb  | 0.600μs | ±0.00% |
| ClassDefinitionBench | benchGetParametersBeforeTransparentCachingMultipleConstructorParameters |     | 5    | 10  | 1.664mb  | 1.600μs | ±0.00% |
+----------------------+-------------------------------------------------------------------------+-----+------+-----+----------+---------+--------+
```

```
laminas-di on  HEAD (79c991a) via 🐘 v8.1.7 via ❄️  impure (nix-shell) 
❯ ./vendor/bin/phpbench report --ref=after_patch_64 --report=aggregate
+----------------------+-------------------------------------------------------------------------+-----+------+-----+----------+---------+--------+
| benchmark            | subject                                                                 | set | revs | its | mem_peak | mode    | rstdev |
+----------------------+-------------------------------------------------------------------------+-----+------+-----+----------+---------+--------+
| ClassDefinitionBench | benchGetParametersBeforeTransparentCachingNoConstructor                 |     | 5    | 10  | 1.664mb  | 0.000μs | ±0.00% |
| ClassDefinitionBench | benchGetParametersBeforeTransparentCachingNoConstructorParameters       |     | 5    | 10  | 1.664mb  | 0.200μs | ±0.00% |
| ClassDefinitionBench | benchGetParametersBeforeTransparentCachingMultipleConstructorParameters |     | 5    | 10  | 1.664mb  | 1.000μs | ±0.00% |
+----------------------+-------------------------------------------------------------------------+-----+------+-----+----------+---------+--------+
```

Seems like a very good improvement, nice! :) 

Obviously, the entire exercise was just to raise awareness of making this a measured improvement, but still, the results are visible, especially in large `laminas/laminas-di` dependency graphs! :+1:

**Comment by Ocramius** [comment]
:ship: will also do a release, thanks @sjokkateer!

**Comment by sjokkateer** [comment]
> ```
> ❯ ./vendor/bin/phpbench run --warmup=2 --revs=5 --iterations=10 --retry-threshold=5 --tag=before_patch_64
> PHPBench (1.2.5) running benchmarks... #standwithukraine
> with configuration file: /home/ocramius/Documents/laminas/laminas-di/phpbench.json
> with PHP version 8.1.7, xdebug ❌, opcache ❌
> 
> \LaminasBench\Di\ClassDefinitionBench
> 
>     benchGetParametersBeforeTransparentCach.R2 I9 - Mo0.200μs (±0.00%)
>     benchGetParametersBeforeTransparentCach.R1 I7 - Mo0.600μs (±0.00%)
>     benchGetParametersBeforeTransparentCach.R2 I7 - Mo1.600μs (±0.00%)
> 
> Subjects: 3, Assertions: 0, Failures: 0, Errors: 0
> Storing results ... OK
> Run: 1348ad5397ac2b4cec5ad3197803445ec0b57c2f
> ```
> 
> ```
> ❯ ./vendor/bin/phpbench run --warmup=2 --revs=5 --iterations=10 --retry-threshold=5 --tag=after_patch_64
> PHPBench (1.2.5) running benchmarks... #standwithukraine
> with configuration file: /home/ocramius/Documents/laminas/laminas-di/phpbench.json
> with PHP version 8.1.7, xdebug ❌, opcache ❌
> 
> \LaminasBench\Di\ClassDefinitionBench
> 
>     benchGetParametersBeforeTransparentCach.R10 I9 - Mo0.000μs (±0.00%)
>     benchGetParametersBeforeTransparentCach.R10 I9 - Mo0.200μs (±0.00%)
>     benchGetParametersBeforeTransparentCach.R2 I2 - Mo1.000μs (±0.00%)
> 
> Subjects: 3, Assertions: 0, Failures: 0, Errors: 0
> 
> Deprecated: DOMDocument::createTextNode(): Passing null to parameter #1 ($data) of type string is deprecated in /home/ocramius/Documents/laminas/laminas-di/vendor/phpbench/dom/lib/Element.php on line 37
> Storing results ... 
> Deprecated: DOMDocument::createTextNode(): Passing null to parameter #1 ($data) of type string is deprecated in /home/ocramius/Documents/laminas/laminas-di/vendor/phpbench/dom/lib/Element.php on line 37
> OK
> Run: 1348ad529b868f7a619cc7024bacef3eb3f71848
> ```
> 
> ```
> laminas-di on  HEAD (79c991a) via 🐘 v8.1.7 via ❄️  impure (nix-shell) 
> ❯ ./vendor/bin/phpbench report --ref=before_patch_64 --ref=after_patch_64 --report=aggregate
> +----------------------+-------------------------------------------------------------------------+-----+------+-----+----------+---------+--------+
> | benchmark            | subject                                                                 | set | revs | its | mem_peak | mode    | rstdev |
> +----------------------+-------------------------------------------------------------------------+-----+------+-----+----------+---------+--------+
> | ClassDefinitionBench | benchGetParametersBeforeTransparentCachingNoConstructor                 |     | 5    | 10  | 1.664mb  | 0.200μs | ±0.00% |
> | ClassDefinitionBench | benchGetParametersBeforeTransparentCachingNoConstructorParameters       |     | 5    | 10  | 1.664mb  | 0.600μs | ±0.00% |
> | ClassDefinitionBench | benchGetParametersBeforeTransparentCachingMultipleConstructorParameters |     | 5    | 10  | 1.664mb  | 1.600μs | ±0.00% |
> +----------------------+-------------------------------------------------------------------------+-----+------+-----+----------+---------+--------+
> ```
> 
> ```
> laminas-di on  HEAD (79c991a) via 🐘 v8.1.7 via ❄️  impure (nix-shell) 
> ❯ ./vendor/bin/phpbench report --ref=after_patch_64 --report=aggregate
> +----------------------+-------------------------------------------------------------------------+-----+------+-----+----------+---------+--------+
> | benchmark            | subject                                                                 | set | revs | its | mem_peak | mode    | rstdev |
> +----------------------+-------------------------------------------------------------------------+-----+------+-----+----------+---------+--------+
> | ClassDefinitionBench | benchGetParametersBeforeTransparentCachingNoConstructor                 |     | 5    | 10  | 1.664mb  | 0.000μs | ±0.00% |
> | ClassDefinitionBench | benchGetParametersBeforeTransparentCachingNoConstructorParameters       |     | 5    | 10  | 1.664mb  | 0.200μs | ±0.00% |
> | ClassDefinitionBench | benchGetParametersBeforeTransparentCachingMultipleConstructorParameters |     | 5    | 10  | 1.664mb  | 1.000μs | ±0.00% |
> +----------------------+-------------------------------------------------------------------------+-----+------+-----+----------+---------+--------+
> ```
> 
> Seems like a very good improvement, nice! :)
> 
> Obviously, the entire exercise was just to raise awareness of making this a measured improvement, but still, the results are visible, especially in large `laminas/laminas-di` dependency graphs! 👍

@Ocramius Glad I can contribute to you guys, also thank you for giving me the opportunity to learn about PHPBench!

**Pull request merged**